`formatjson.ps1` - Sort manifest root keys by `schema.json` and child keys alphabetically

o-l-a-v opened the pull request:
#### Description

1. Reordered root properties in `schema.json` so it matches the wanted order.
1. Created new functions in `manifest.ps1`
   * `Sort-ScoopManifestProperties`:
       * Sorts root properties according to `schema.json` properties.
       * Sorts one level deep child properties alphabetically if the parent property is of type `PSCustomObject`.
       * Called by `formatjson.ps1`.
   * `Sort-PSCustomObjectKeysRecursively`:
      * Uses recursion to sort all child keys alphabetically, does not touch other values.
      * Called by `Sort-ScoopManifestProperties` for manifest 1 level child properties that are of type `[PSCustomObject]`.
1. Made `formatjson.ps1` use new function `Sort-ScoopManifestProperties`.

#### Motivation and Context

Closes #6491.

#### How Has This Been Tested?

```pwsh
# Path to formatjson.ps1 in branch "formatjson-sort-root-properties"
$FunctionPath = [string] 'D:\git\ScoopInstaller--Scoop\bin\formatjson.ps1'

# Path to main bucket in branch "master" pulled 2025-09-19T12:25:00+02:00
$Directory = [string] 'D:\git\ScoopInstaller--Main\bucket'

# Run format json, look at git afterwards
$ErrorActionPreference = 'Stop'
. $FunctionPath -Dir $Directory -App '*'
```

<img width="2551" height="763" alt="image" src="https://github.com/user-attachments/assets/da230491-926c-464a-b914-66cc4676eb32" />

#### Checklist:

- [x] I have read the [Contributing Guide](https://github.com/ScoopInstaller/.github/blob/main/.github/CONTRIBUTING.md).
- [x] I have ensured that I am targeting the `develop` branch.
- [ ] I have updated the documentation accordingly.
   * Info about `formatjson.ps1` was added with <https://github.com/ScoopInstaller/GithubActions/pull/61>: <https://github.com/ScoopInstaller/GithubActions/wiki/Pull-Request-Checks>
- [ ] I have updated the tests accordingly.
- [x] I have added an entry in the CHANGELOG.


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Manifest formatter now deterministically orders root and first-level properties; added two manifest utilities to support recursive and root-level ordering.
  * Schema expanded with new top-level fields and updated required fields (version, homepage, license).

* **Refactor**
  * More robust manifest parsing and safer formatting pipeline that writes back only when output is valid.

* **Tests**
  * Added test to ensure empty-path JSON parsing does not throw.

* **Documentation**
  * CHANGELOG updated.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

o-l-a-v committed:
Reorder properties to wanted order

o-l-a-v committed:
Add sort function to json.ps1

o-l-a-v committed:
Take PSCustomObject instead, let parse_json do the parsing

o-l-a-v committed:
Added sorting manifest as a step

o-l-a-v committed:
Make variable name more telling

o-l-a-v committed:
Added changes to changelog

o-l-a-v committed:
Don't use aliases

o-l-a-v committed:
Use existing function parse_json instead

o-l-a-v committed:
Make parse_json function more readable

o-l-a-v committed:
Fix goof by me

o-l-a-v committed:
Read and parse schema.json once

o-l-a-v commented:
@coderabbitai review

z-Fng commented on bin/formatjson.ps1:
```diff
@@ -34,11 +34,21 @@ param(
+    $json = [string]($json -replace "`t", '    ')
+
+    # Overwrite file content
     [System.IO.File]::WriteAllLines($file, $json)
```
When JSON parsing fails, it will clear the manifest content. Should this behavior be adjusted to prevent contributors from losing their work?

o-l-a-v replied:
Yup, maybe we should add some error handling.

z-Fng commented on bin/formatjson.ps1:
```diff
@@ -34,11 +34,21 @@ param(
+    $json = [PSCustomObject](parse_json -path $file)
+
+    # Sort JSON root properties
+    $json = [PSCustomObject](Sort-ScoopManifestRootProperties -JsonAsObject $json)
```
Is it possible to sort the keys recursively with global/local order?

o-l-a-v replied:
I don't fully understand what you're asking for here, about the local/global order.

Recursice ordering keys by the schema.json should be possible. Or just sort child keys alphabetically.

z-Fng replied:
> local/global order

My description might be redundant. Jsut ignore it. Focus on recursive sorting.

I thought that when sorting keys like architecture.64bit/32bit , local order are used because the configuration file contains architecture.64bit/32bit. However, when sorting keys like architecture.64bit.url/hash, the configuration file lacks this entry, causing it to fall back to the global order.

As I'm not sure how to describe this, I called it  local/global order.

https://github.com/ScoopInstaller/Scoop/blob/37aa4596d5f1e5b8d914786b6f038e597ac52960/schema.json#L217-L231

o-l-a-v replied:
I added sorting of first level child keys if parent is of type `PSCustomObject`. Going even deeper will be more complex. It's doable, but I think what we got now is a good start.

o-l-a-v replied:
Maybe there are existing PowerShell or C# projects that can sort a JSON file by `schema.json` and handle all the complexity of going deeper we can use?

In Python there is: <https://github.com/ikonst/jschon-sort>.

But as I already said, I think this serves as a very good start for getting more uniform manifests.

z-Fng replied:
> Going even deeper will be more complex. It's doable, but I think what we got now is a good start.

Yes. It is a good start. The changes made so far are getting a bit complex. Thankfully, it's just a tool.

> Maybe there are existing PowerShell or C# projects that can sort a JSON file by schema.json and handle all the complexity of going deeper we can use?

Unfortunately, I haven't been able to find any yet.

z-Fng commented on bin/formatjson.ps1:
```diff
@@ -34,11 +34,21 @@ param(
+    $json = [PSCustomObject](Sort-ScoopManifestRootProperties -JsonAsObject $json)
+
+    # Beautify
+    $json = [string](ConvertToPrettyJson -data $json)
```
I am concerned whether the sort is a stable sort algorithm for keys that are not present in the index list.

Could multiple executions yield different results?

o-l-a-v replied:
IMO there could and should be a check for keys we don't want to be present. Doesn't this exist already?

To make formatjson.ps1 safer we could do: If key is not present in the schema.json

* Optian a) fail execution
* Option b) remove key

z-Fng replied:
> Doesn't this exist already?

Does it? It will fail execution currently.
```
Sort-Object: D:\Scoop\apps\scoop\current\lib\json.ps1:241:9
Line |
 241 |          Sort-Object -Property @{
     |          ~~~~~~~~~~~~~~~~~~~~~~~~
     | Cannot convert value "-1" to type "System.Byte". Error: "Value was either too large or too small for an unsigned byte."
```

o-l-a-v replied:
I was thinking more about that other parts of the validation should catch manifest errors before it gets to `formatjson.ps1`. Thus validating that the manifest doesn't contain invalid keys is a task that `formatjson.ps1` shouldn't try to solve.

o-l-a-v replied:
Seems that's what the validator is for?

<https://github.com/ScoopInstaller/Scoop/tree/master/supporting/validator>

Thus I don't think the new function `Sort-ScoopManifestRootProperties` should have a lot of validation logic. It's a feature that it throws if manifest does not comply with `schema.json` IMO.

o-l-a-v replied:
I added a check for parent keys not present in `schema.json` inside `Sort-ScoopManifestRootProperties`. Then users will get a better error message than what you posted in <https://github.com/ScoopInstaller/Scoop/pull/6494#discussion_r2365560938>.

o-l-a-v replied:
Or we could write a warning, and just return the object as is / not try to sort the object at all. Instead of throwing.

z-Fng replied:
> Seems that's what the validator is for?

Yes, thanks for reminding.


> I added a check for parent keys not present in `schema.json` inside `Sort-ScoopManifestRootProperties`. Then users will get a better error message than what you posted.

Totally agree. Great.

o-l-a-v committed:
Add some failproof to make sure only expected keys are present

o-l-a-v committed:
Remove debug

o-l-a-v committed:
Sort childs alphabetically

o-l-a-v committed:
Only overwrite JSON file if new content is not empty

o-l-a-v committed:
Rename function and add synopsis

o-l-a-v committed:
Use Scoop function abort instead of PowerShell native Throw

z-Fng commented:
@coderabbitai review

o-l-a-v committed:
Updated changelog to correct name of new function

o-l-a-v committed:
Use System.IO.File WriteAllText, because ConvertToPrettyJson returns string, not string array.

o-l-a-v committed:
WriteAllText does not write the final newline, so reverting back to WriteAllLines

o-l-a-v committed:
Use uin16 instead of byte, just in case Scoop manifest ever gets more than 255 properties

o-l-a-v committed:
Validate that input is not null or empty

o-l-a-v commented on lib/json.ps1:
```diff
@@ -212,3 +212,65 @@ function normalize_values([psobject] $json) {
+    Param(
+        [Parameter(Mandatory)]
+        [PSCustomObject] $JsonAsObject
+    )
```
Added validation as `[ValidateScript({})]`.

o-l-a-v commented:
I implemented some of the nitpicks from CodeRabbit.

* CHANGELOG: Fixed.

> bin/formatjson.ps1 (2): Use WriteAllText for a single string payload

I tested this, this does not write the final newline. So I kept it to use WriteAllLines.

> lib/json.ps1: Cache wanted order; avoid hidden dependency on parse_json

Scoop does this many other places, uses it's own functions. I can change this if maintainers want it, but that would not fit with what existing code already does.

> Avoid [byte] cast; use int to prevent overflow and past -1 issues

It's highly unlikely that the Scoop manifest will ever get more than 255 unique root properties. But I changed it to uint16 just in case, which will do 65000++.

o-l-a-v committed:
Moved function to manifest.ps1 + made sorting of child keys recursive

o-l-a-v committed:
Apply nitpicks from CodeRabbit, except the case of handling arrays as we don't want that.

o-l-a-v committed:
Apply one more nitpick

o-l-a-v commented on lib/manifest.ps1:
```diff
@@ -212,3 +218,106 @@ function uninstaller($manifest, $arch) { arch_specific 'uninstaller' $manifest $
+    )
+
+    # Get wanted order from Scoop manifest schema
+    if ([string]::IsNullOrWhiteSpace($Script:WantedOrder)) {
+        $Script:WantedOrder = [string[]](
+            (
+                parse_json -path (
+                    '{0}\..\schema.json' -f $PSScriptRoot
+                )
+            ).'properties'.'PSObject'.'Properties'.'Name'
+        )
+    }
```
> IndexOf on a string[] isn’t an instance method; use [Array]::IndexOf(...)

This is not true. `[string[]]` also has the `.IndexOf()` method.

```pwsh
PS > $([string[]]('0','1','2')).IndexOf('1')
1
PS >
```

z-Fng commented on lib/manifest.ps1:
```diff
@@ -2,10 +2,16 @@ function manifest_path($app, $bucket) {
-    if ($null -eq $path -or !(Test-Path $path)) { return $null }
+function parse_json {
+    Param(
+        [Parameter(Mandatory)]
```
Adding `[Parameter(Mandatory)]` changes the original logic. I encountered the following error while testing.
```console
> scoop download uninstalled-app
WARNING: Searching manifest: uninstalled-app, main
parse_json: D:\Scoop\apps\scoop\current\lib\manifest.ps1:123
Line |
 123 |      parse_json (manifest_path $app $bucket)
     |                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | Cannot bind argument to parameter 'path' because it is an empty string.
WARNING: Searching manifest: uninstalled-app, extras
parse_json: D:\Scoop\apps\scoop\current\lib\manifest.ps1:123
Line |
 123 |      parse_json (manifest_path $app $bucket)
     |                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | Cannot bind argument to parameter 'path' because it is an empty string.
```

o-l-a-v replied:
Good find, I'll revert.

Can this be added as an automated test?

z-Fng replied:
Scoop seems to support unit tests only and does not support integration tests currently.

Perhaps we can add tests for `parse_json` in the unit tests?

o-l-a-v committed:
Add test for parse_json and an empty string

o-l-a-v committed:
Merge branch 'develop' into formatjson-sort-root-properties

o-l-a-v committed:
Make path not required

z-Fng commented:
> PSCustomObject.properties won't output anything

It's actually 'properties' in schema.json.

> Is it LLM generated?

Yes, after editing and modifying it, it runs fine on my local machine.
